test/integration/self-provider-staking.test.ts

```diff
@@ -0,0 +1,586 @@
+      it('failure - wrong address format', async () => {
+        const res = provider.contract.getUserStakes(`0xinvalid`)
+
+        await expect(res).rejects.toThrow(commonErrorMessage.wrongAddressFormat)
+      })
+
+      it('success - get user total stakes', async () => {
+        const res = await provider.contract.getUserStakes(signerAddress)
+
+        expect(Array.isArray(res)).toBe(true)
+      })
+    })
+
+    describe('setDates()', () => {
+      it('failure - return error if not the admin calls this function', async () => {
+        const res = notAdminProvider.contract.setDates(1609545600, 1609459200)
+
+        await expect(res).rejects.toThrow(commonErrorMessage.addressIsNotAdmin)
+      })
+
+      it('failure - staking end time < rewards end time', async () => {
+        await expect(
+          provider.contract.setDates(1609459200, 1609545600)
+        ).rejects.toThrow(stakingErrorMessage.stakingCantBeLessThanRewardsEnd)
+      })
+
+      it.skip('success - set dates successfully', async () => {
+        await expect(
+          provider.contract.setDates(1609545600, 1609459200)
+        ).resolves.not.toThrow()
+      })
+    })
+
+    describe('withdraw()', () => {
+      it('failure - return error if not the admin calls this function', async () => {
+        const res = notAdminProvider.contract.withdraw(
+          FACTR_TOKEN_ADDRESS,
+          '0xinvalid'
+        )
+
+        await expect(res).rejects.toThrow(commonErrorMessage.addressIsNotAdmin)
+      })
+
+      it('failure - invalid token address format', async () => {
+        await expect(
+          provider.contract.withdraw('0xinvalid', '0xinvalid')
+        ).rejects.toThrow(commonErrorMessage.wrongAddressFormat)
+      })
+
+      it('failure - invalid recipient address format', async () => {
+        await expect(
+          provider.contract.withdraw(FACTR_TOKEN_ADDRESS, '0xinvalid')
+        ).rejects.toThrow(commonErrorMessage.wrongAddressFormat)
+      })
+
+      it.skip('success - withdraw successfully', async () => {
+        await expect(
+          provider.contract.withdraw(FACTR_TOKEN_ADDRESS, signerAddress)
+        ).resolves.not.toThrow()
+      })
+    })
+
+    describe('claimRewards()', () => {
+      it('failure - contract is paused', async () => {
+        const tx = await provider.contract.pause()
+
+        await waitUntilConfirmationCompleted(
+          provider.contract.jsonRpcProvider,
+          tx
+        )
+
+        await expect(provider.contract.claimRewards(BigInt(0))).rejects.toThrow(
+          commonErrorMessage.contractIsPaused
+        )
+
+        const tx2 = await provider.contract.unpause()
+
+        await waitUntilConfirmationCompleted(
+          provider.contract.jsonRpcProvider,
+          tx2
+        )
+      })
+
+      it('failure - stake is already unstaked', async () => {
+        const stakeIndex = BigInt(0)
+        // const unstakeTx = await provider.contract.unstake(stakeIndex)
```

Comment:
Is this block comment necessary?

Reply:
I decided to left it here as a reference for when we have a local testing environment, this commented code would be required to create an stakes to make the full test.